Integrate external api

wleoguty91 opened the pull request:
## Summary

- Integrated with external API

## Trello Card

- https://trello.com/c/F6PXTkBs/22-integraci%C3%B3n-api-externa

wleoguty91 committed:
Integrate external api

wleoguty91 committed:
Reformat code style

leonplondon commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -15,15 +19,26 @@
+
+    @Bean
+    private OpenLibraryService openLibraryService() {
+        return new OpenLibraryService();
```
Maybe those beans should be in a class annotated with `@Configuration`

leonplondon commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -35,24 +50,38 @@ public Book create(@RequestBody Book book) {
-    public Book findOne(@PathVariable Long id) {
-        return bookRepository.findById(id).orElseThrow(BootstrapMethodError::new);
+    public Book findOne(@PathVariable Long id) throws IOException {
+        return bookRepository.findById(id).orElseThrow(this::bookNotFound);
```
Maybe this exception can be handled with a custom one with a more information (more readable) not throwing it at the framework level

leonplondon commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -35,24 +50,38 @@ public Book create(@RequestBody Book book) {
+
+    @GetMapping("/search")
+    public ResponseEntity<Book> search(@RequestParam(name = "isbn", required = false) String isbn)
+        throws IOException {
```
What does happen if isbn is not provided?

wleoguty91 replied:
This happend!

![Screenshot from 2020-01-02 16-27-03](https://user-images.githubusercontent.com/42581484/71687508-c2771480-2d7c-11ea-863c-ddbd8c46937d.png)

leonplondon commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,71 @@
+        Optional<BookDTO> optionalBookDTO = bookInfo(isbn);
+        return optionalBookDTO.map((bookDTO) -> {
+            Book book = convertDtoToEntity(bookDTO);
+            bookRepository.save(book);
```
Please return the newly created book, the one returned by `repository.save`

leonplondon commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public Optional<BookDTO> bookInfo(String isbn) throws IOException {
+        RestTemplate restTemplate = new RestTemplate();
```
Maybe this could be a dependency that must be injected not instantiated.

wleoguty91 committed:
Move exceptions and configurations
Move bean of openlibraryservice to a config file, added
exceptions for book not found and book id dont match

facucompe commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,71 @@
+import wolox.training.models.Book;
+import wolox.training.repositories.BookRepository;
+
+public class OpenLibraryService {
```
Make this class a service bean adding `@Service` at the top

facucompe commented on src/main/java/wolox/training/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,14 @@
+public class ApplicationConfig {
+
+    @Bean
+    public OpenLibraryService openLibraryService() {
```
If you make OpenLibraryService a service bean, you don't have to do this. Remove this configuration

facucompe commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,71 @@
+
+    public Optional<BookDTO> bookInfo(String isbn) throws IOException {
+        RestTemplate restTemplate = new RestTemplate();
+        String urlString = "https://openlibrary.org/api/books?bibkeys=";
```
Create a property for thi URL

leonplondon commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -12,21 +15,28 @@
+    @Autowired
+    private OpenLibraryService openLibraryService;
+
     public BookController(BookRepository bookRepository) {
```
Same as previous PR, using annotation.

wleoguty91 committed:
Change open library to service
Changed the config of openlibraryservice to a service prop, also
deleted the application config, move baseurl to a config file

leonplondon commented on src/main/java/wolox/training/exceptions/BookIdDontMatchException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_ACCEPTABLE, reason = "Book id don't match")
+public class BookIdDontMatchException extends Exception {
```
You could extend from `RuntimeException` so you avoid writing logicless constructor and cleaning controllers from re-throwing exceptions.

leonplondon commented on src/main/java/wolox/training/controllers/BookController.java:
```diff
@@ -28,8 +33,13 @@ public BookController(BookRepository bookRepository) {
     }
 
+    @Autowired
+    private OpenLibraryService openLibraryService;
```
This should be before controller's constructor and also remember to inject dependencies through the constructor not at field level.

leonplondon commented on src/main/java/wolox/training/dto/BookDTO.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Data;
+
+@Data
+public class BookDTO implements Serializable {
```
Why do you need to implement "Serializable" interface?

wleoguty91 replied:
Because I took this [example](https://www.oscarblancarteblog.com/2018/11/30/data-transfer-object-dto-patron-diseno/), that use this interface... But is not necessary, I can remove it

facucompe commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,76 @@
+        RestTemplate restTemplate = new RestTemplate();
+        String urlString = baseUrl + String.format("?bibkeys=ISBN:%s&format=json&jscmd=data", isbn);
+        ResponseEntity<String> response = restTemplate.getForEntity(urlString, String.class);
+        ObjectMapper mapper = new ObjectMapper();
```
You can use a `@PostConstruct`  to instantiate this object

facucompe commented on src/main/java/wolox/training/services/OpenLibraryService.java:
```diff
@@ -0,0 +1,76 @@
+        return String.join(", ", root.findValuesAsText(key));
+    }
+
+    private Book convertDtoToEntity(BookDTO bookDTO) {
```
Seems like you can map the response directly to the model

leonplondon reviewed (dismissed):
Some comments of this PR were left to take into account on future PRs

wleoguty91 committed:
Refactors by cr suggestions
